components/raftstore-v2/src/operation/query/replica_read.rs

```
@@ -0,0 +1,111 @@
+        read_index_req: &mut ReadIndexRequest<QueryResChannel>,
+        ctx: &mut StoreContext<EK, ER, T>,
+    ) {
+        debug!(
+            self.logger,
+            "handle replica reads with a read index";
+            "request_id" => ?read_index_req.id,
+        );
+        RAFT_READ_INDEX_PENDING_COUNT.sub(read_index_req.cmds().len() as i64);
+        for (req, ch, mut read_index) in read_index_req.take_cmds().drain(..) {
+            // leader reports key is locked
+            if let Some(locked) = read_index_req.locked.take() {
+                let mut response = raft_cmdpb::Response::default();
+                response.mut_read_index().set_locked(*locked);
+                let mut cmd_resp = RaftCmdResponse::default();
+                cmd_resp.mut_responses().push(response);
+                ch.set_result(QueryResult::Response(cmd_resp));
+                continue;
+            }
+            if req.get_header().get_replica_read() {
+                // We should check epoch since the range could be changed.
+                ch.set_result(self.can_replica_read(req, true, read_index_req.read_index));
+            } else {
+                // The request could be proposed when the peer was leader.
+                // TODO: figure out that it's necessary to notify stale or not.
+                let term = self.term();
+                notify_stale_req(term, ch);
+            }
+        }
+    }
+
+    pub(crate) fn can_replica_read(
+        &self,
+        req: RaftCmdRequest,
+        check_epoch: bool,
+        read_index: Option<u64>,
+    ) -> QueryResult {
+        let region = self.region().clone();
+        if check_epoch {
+            if let Err(e) = check_region_epoch(&req, &region, true) {
+                debug!(self.logger, "epoch not match"; "err" => ?e);
+                let mut response = cmd_resp::new_error(e);
+                cmd_resp::bind_term(&mut response, self.term());
+                return QueryResult::Response(response);
```

This is actually a mis-usage. `QueryResult::Response` is for status command. In this case, it should return an error by `ch.report_error(resp)`.